rfc: Fuzz and Auto Debug Tool (#125)
pingcap/community: rfc/2019-12-07-fuzz-and-auto-debug.md

# Proposal: Fuzz and Auto Debug Tool

- Author(s):     [DQinYuan](https://github.com/DQinYuan)
- Last updated:  2019-12-07
- Discussion at:  [issue#124](https://github.com/pingcap/community/issues/124)

## Abstract

> There is a lot of uncertainty in Chaos and Fuzz test of complicated software like TiDB. Developers often costs a few of weeks to localize the root cause. This proposal may help on this with fuzz mutation and auto debug.

Reviewer: ```suggestion
There is a lot of uncertainty in Chaos and Fuzz test of complicated software like TiDB. Developers often cost a few of weeks to locate the root cause. This proposal may help to improve this with fuzz mutation and auto debug.
```

> 

Reviewer: 1 empty line is enough.

Reviewer: So as the other places.

Author: OK, I have fixed it.

## Background

> There are many troublesome bugs,  you don't know what's going wrong, even worse, you can not reproduce it .  

Reviewer: ```suggestion
There are many troublesome bugs, you don't know what's going wrong, even worse, you can not reproduce it .  
```

> [Statistical debug](https://stackoverflow.com/questions/505907/what-is-statistical-debugging) is a common auto debug idea, but it need a lot of test cases to accurately localize bug position.

Reviewer: ```suggestion
[Statistical debug](https://stackoverflow.com/questions/505907/what-is-statistical-debugging) is a common auto debug idea, but it needs a lot of test cases to accurately locate the bug position.
```

> Fuzz test is a good idea to apply unlimited case to application, it will provide a bulk of cases to Statistical Debug.  Mutation-Based Fuzzing can introduce small changes  to existing inputs that may still keep the input valid. Some paper guide mutations with some  algorithms to converge input in a root bug. 

Reviewer: ```suggestion
Fuzz test is a good idea to apply unlimited cases to application, it will provide a bulk of cases to Statistical Debug. Mutation-Based Fuzzing can introduce small changes to existing inputs that may still keep the input valid. Some paper guide mutations with some algorithms to converge input in a root bug. 
```

## Proposal

> This tool will archieve following goals:

Reviewer: ```suggestion
This tool achieves the following goals:
```

- SQLFuzz with mutation
- General fuzz with mutation
- Auto debug in Go and Rust code by highlighting suspicious code block

## Rationale

> We already test TiDB with some simple fuzz test tool, like randgen or sqlsmith, which only do random fuzz，without direction to converge . Some fuzz tool with mutation can   mutate input  from some seeds gradually, like [libFuzzer](https://llvm.org/docs/LibFuzzer.html) or [Angora](https://github.com/AngoraFuzzer/Angora).  We can guide fuzz mutation by some algorithms（for example, [Angora](https://github.com/AngoraFuzzer/Angora) is guided by branch coverage; [NEZHA](http://www.cs.columbia.edu/~suman/docs/nezha.pdf) is guided by  delta-diversity ） ,  so that it will converge to some bug. We will apply this technology in more general area,  not only apply on SQLFuzz, but also in general fuzz.

Reviewer: please remove the unnecessary spaces between words, we only need 1 space.

Author: OK, I have fixed it

check pass or not  |            |         |                                     |                      |
                   |            | +-------+-+       +-----------------+         |                      |
                   +------------> |collector+------>+statistical model+--------------------------------+
                                | +---------+       +-------+---------+         |
                                |                           ^                   |
                                |                           |                   |
                                +-----------------------------------------------+
                                                            |
                                                            |
                                                       +----+---+
                                                       |frontend|
                                                       +--------+

```

### Instruction tool

Like [go tool cover](), we will add code at the beginning of every code basic block in conpiler phase:

```go

>         trace_util_0.Count(_main_00000, 0)   // Instruction

Reviewer: please remove the prefix spaces.

Author: OK, I have fixed it

Reviewer: The prefix spaces still exists, I'm going to reopen this conversion.

Author: Do you mean prefix spaces in code? I mistake it to the empty line behind the code block...

        flag.Parse()
        if *version {
                trace_util_0.Count(_main_00000, 5)  // Instruction
                fmt.Println(printer.GetTiDBInfo())
                os.Exit(0)
        }
        trace_util_0.Count(_main_00000, 1)   // Instruction
        registerStores()
        //.....
```

`trace_util_0` in above code is an extra package generated by Instruction tool, this package will record start and end line number of every basic block:

```go

>         pathCounter := &CoverCounter{

Reviewer: ditto

Author: OK, I have fixed it

Reviewer: The prefix spaces still exists, I'm going to unresolve this conversion.

Author: I mistake before. Now I know what you mean.
